keps/sig-node/4639-oci-volume-source/README.md

```diff
@@ -0,0 +1,1391 @@
+which is the internal API compared to the external one from staging. The [API Validation](https://github.com/kubernetes/kubernetes/blob/7b359a2f9e1ff5cdc49cfcc4e350e9d796f502c0/pkg/apis/core/validation/validation.go)
+validation will be extended to disallow the `subPath`/`subPathExpr` field as
+well as making the `reference` mandatory:
+
+```go
+// …
+
+if source.OCI != nil {
+	if numVolumes > 0 {
+		allErrs = append(allErrs, field.Forbidden(fldPath.Child("oci"), "may not specify more than 1 volume type"))
+	} else {
+		numVolumes++
+		allErrs = append(allErrs, validateOCIVolumeSource(source.OCI, fldPath.Child("oci"))...)
+	}
+}
+
+// …
+```
+
+```go
+func validateOCIVolumeSource(oci *core.OCIVolumeSource, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	if len(oci.Reference) == 0 {
+		allErrs = append(allErrs, field.Required(fldPath.Child("reference"), ""))
+	}
+	allErrs = append(allErrs, validatePullPolicy(oci.PullPolicy, fldPath.Child("pullPolicy"))...)
+	return allErrs
+}
+```
+
+```go
+// …
+
+// Disallow subPath/subPathExpr for OCI volumes
+if v, ok := volumes[mnt.Name]; ok && v.OCI != nil {
+	if mnt.SubPath != "" {
+		allErrs = append(allErrs, field.Invalid(idxPath.Child("subPath"), mnt.SubPath, "not allowed in OCI volume sources"))
+	}
+	if mnt.SubPathExpr != "" {
+		allErrs = append(allErrs, field.Invalid(idxPath.Child("subPathExpr"), mnt.SubPathExpr, "not allowed in OCI volume sources"))
+	}
+}
+
+// …
+```
+
+### Kubelet and Container Runtime Interface (CRI) support for OCI artifacts
+
+Kubelet and the Container Runtime Interface (CRI) currently handle OCI images. To support OCI artifacts,
+potential enhancements may be required:
+
+**Extended Media Type Handling in the container runtime:**
+   - Update container runtimes to recognize and handle new media types associated with OCI artifacts.
+   - Ensure that pulling and storing these artifacts is as efficient and secure as with OCI images.
+
+**Lifecycling and Garbage Collection:**
+   - Reuse the existing kubelet logic for managing the lifecycle of OCI objects.
+   - Extending the existing image garbage collection to not remove an OCI volume
+     image if a pod is still referencing it. 
+
+**Artifact-Specific Configuration:**
+   - Introduce new configuration options to handle the unique requirements of different types of OCI artifacts.
+
+**Artifacts as Subject Referrers:**
+   - Introduce new refer to image and filter for artifact type criterion/options for to be mounted artifact(s).
+   - Certain types of OCI artifacts include a subject reference. That reference
+     identifies the artifact/image for which this artifact refers. For example a
+     signature artifact could refer to a platform index, for certifying the
+     platform images, or to an SBOM artifact that refers to a platform matched
+     image. These artifacts may or may not be located on the same
+     registry/repository. The new referrers API allows for retrieving these
+     artifacts.
```

Comment:
While it is possible the subject could reference content in a different repository, registry behavior may vary (a GC could delete the data as an untagged manifest), and the referrers API will only respond with content found in the requested repository.

```suggestion
     image. These artifacts may or may not be located on the same
     registry/repository. The new referrers API allows for discovering
     artifacts from a requested repository.
```

Reply:
```suggestion
     image. These artifacts may or may not be located on the same
     registry/repository. The new referrers API allows for discovering subject referring artifacts in an OCI v1.1 compatible registry. These subject referring artifacts may then be downloaded for further processing at the client.
```

Reply:
@mikebrow I believe your edit leaves the false impression that the referrers API would respond with artifacts pushed to a different repository or registry. Is that intentional? If so, lets discuss. Perhaps some registries could be implemented without the concept of repository namespacing and always respond to requests with a global namespace regardless of the specified repository. But if they exist, documenting them here would cause a lot of confusion for end users.

Is "subject referring artifacts" a new term that needs to be defined?

I'd also avoid the v1.1 mention. The API was added in v1.1, but we have a fallback process documented in the spec that will work for older registries and could be used by tools implementing this.

Reply:
(I added the suggestion from @sudo-bmitch)

Reply:
was not intended to indicate registry `a` knew anything about artifacts in registry `b`  So yeah no, didn't mean to give a false impression.. Would've said all or any instead of using the `a` singular.  

Don't think a new term is needed.. an artifact that has a subject reference and is thus discoverable by referrers seems to be a subject referring artifact, separate from artifacts that do not have a subject reference field.

Of course for others reading, the fall back process, also defined in 1.1, does not use the referrers api, rather a pushed index manifest, and this is the referrers section of this doc, suppose we could change it to referrers and/or the backup processes using an index with a list of artifacts or other means. But yes we could mention other techniques other than using referrers to retrieve artifacts.

Reply:
> suppose we could change it to referrers and/or the backup processes using an index with a list of artifacts or other means

I should clarify my 1.1 comment. I'd be fine with a call out to the "OCI 1.1 distribution-spec", perhaps as a link to the referrers implementation. I just didn't want to imply a requirement on a "1.1 compatible registry" for a client implementation to be able to work.